compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -150,97 +147,198 @@ generateSrcFromLf thisPkgId pkgMap m = mkNoLoc mod
             { hsmodImports = imports
             , hsmodName =
                   Just
-                      (mkNoLoc $
+                      (noLoc $
                        mkModuleName $
                        T.unpack $ LF.moduleNameString $ LF.moduleName m)
             , hsmodDecls = decls
             , hsmodDeprecMessage = Nothing
             , hsmodHaddockModHeader = Nothing
             , hsmodExports = Nothing
             }
+    templateTy =
+        noLoc $
+        HsTyVar NoExt NotPromoted $
+        noLoc $
+        mkRdrQual (mkModuleName "DA.Internal.Template") $
+        mkOccName varName "Template" :: LHsType GhcPs
+    sigRdrName = noLoc $ mkRdrUnqual $ mkOccName varName "signatory"
     decls =
-        [ mkNoLoc $
-        TyClD NoExt $
-        DataDecl
-            { tcdDExt = NoExt
-            , tcdLName =
-                  mkNoLoc $
-                  mkRdrUnqual $
-                  mkOccName varName $
-                  T.unpack $
-                  sanitize $ T.intercalate "." $ LF.unTypeConName dataTypeCon
-            , tcdTyVars =
-                  HsQTvs
-                      { hsq_ext = NoExt
-                      , hsq_explicit =
-                            [ mkUserTyVar $ LF.unTypeVarName tyVarName
-                            | (tyVarName, _kind) <- dataParams
-                            ]
-                      }
-            , tcdFixity = Prefix
-            , tcdDataDefn =
-                  HsDataDefn
-                      { dd_ext = NoExt
-                      , dd_ND = DataType
-                      , dd_ctxt = mkNoLoc []
-                      , dd_cType = Nothing
-                      , dd_kindSig = Nothing
-                      , dd_cons = convDataCons dataTypeCon dataCons
-                      , dd_derivs = mkNoLoc []
-                      }
-            }
-        | LF.DefDataType {..} <- NM.toList $ LF.moduleDataTypes m
-        ]
+        concat $ do
+            LF.DefDataType {..} <- NM.toList $ LF.moduleDataTypes m
+            let templType =
+                    LF.mkTApps
+                        (LF.TCon
+                             (LF.Qualified LF.PRSelf (LF.moduleName m) dataTypeCon))
+                        (map (LF.TVar . fst) dataParams)
+            let occName =
+                    mkOccName varName $
+                    T.unpack $
+                    sanitize $ T.intercalate "." $ LF.unTypeConName dataTypeCon
```

Comment:
We should fail hard if the type constructor name has more than one component, unless it's coming from a sum-of-products type.

Reply:
I'll do that in the next PR, where I fix the sum of products translation.

Reply:
Makes sense.